Add operator logic (#20)
ERC4626-Alliance/ERCs: ERCS/erc-7540.md

      type: uint256
```

#### `isOperator`

Returns `true` if the `operator` is approved as an operator for an `owner`.

```yaml
- name: isOperator
  type: function
  stateMutability: view

  inputs:
    - name: owner
      type: address
    - name: operator
      type: address

  outputs:
    - name: status
      type: bool
```

#### `setOperator`

Grants or revokes permissions for `operator` to manage Requests on behalf of the `msg.sender`.

MUST set the operator status to the `approved` value.

MUST log the `OperatorSet` event.

MUST return True.

```yaml
- name: setOperator
  type: function
  stateMutability: nonpayable

  inputs:
    - name: operator
      type: address
    - name: approved
      type: bool

  outputs:
    - name: success
      type: bool
```

#### `deposit` and `mint` overloaded methods

> Implementations MUST support overloaded `deposit` and `mint` methods based on the specification from [ERC-4626](./eip-4626.md), with an additional `owner` input of type `address`.

Reviewer: ```suggestion
Implementations MUST support an additional overloaded `deposit` and `mint` method on the specification from [ERC-4626](./eip-4626.md), with an additional `owner` input of type `address`:

- `deposit(uint256 assets, address receiver, address owner)`
- `mint(uint256 shares, address receiver, address owner)`
```

      type: uint256
```

#### `isOperator`

Returns `true` if the `operator` is approved as an operator for an `owner`.

```yaml
- name: isOperator
  type: function
  stateMutability: view

  inputs:
    - name: owner
      type: address
    - name: operator
      type: address

  outputs:
    - name: status
      type: bool
```

#### `setOperator`

> Grants or revokes permissions for `operator` to manage Requests on behalf of the `msg.sender`.

Reviewer 2: should this be: "Grants or revokes permissions for `operator` to manage Requests on behalf of the `owner`"

#### `isOperator`

Returns `true` if the `operator` is approved as an operator for an `owner`.

```yaml
- name: isOperator
  type: function
  stateMutability: view

  inputs:
    - name: owner
      type: address
    - name: operator
      type: address

  outputs:
    - name: status
      type: bool
```

#### `setOperator`

Grants or revokes permissions for `operator` to manage Requests on behalf of the `msg.sender`.

MUST set the operator status to the `approved` value.

MUST log the `OperatorSet` event.

MUST return True.

```yaml
- name: setOperator
  type: function
  stateMutability: nonpayable

  inputs:
    - name: operator
      type: address
    - name: approved
      type: bool

  outputs:
    - name: success
      type: bool
```

#### `deposit` and `mint` overloaded methods

Implementations MUST support an additional overloaded `deposit` and `mint` method on the specification from [ERC-4626](./eip-4626.md), with an additional `owner` input of type `address`:

- `deposit(uint256 assets, address receiver, address owner)`
- `mint(uint256 shares, address receiver, address owner)`

The `owner` field is used to look up the Request for which the `assets` should be claimed.

> `owner` MUST equal `msg.sender` unless the `owner` has approved the `msg.sender` as an operator.

Reviewer: ```suggestion
Calls MUST revert unless `msg.sender` is either equal to `owner` or an operator approved by `owner`.
```

> WHen the `Deposit` event is emitted, the first parameter MUST be the `sender`, and the second parameter MUST be the `owner`.

Reviewer: ```suggestion
When the `Deposit` event is emitted, the first parameter MUST be the `sender`, and the second parameter MUST be the `owner`.
```

      type: uint256
```

#### `OperatorSet`

The `owner` has set the `approved` status to an `operator`.

MUST be logged when the operator status is set.

MAY be logged when the operator status is set to the same status it was before the current call.

```yaml
- name: OperatorSet
  type: event

  inputs:
    - name: owner
      indexed: true
      type: address

>     - name: operator

Reviewer: Why not keep the naming consistent with ERC6909 and change operator to spender

Author: Spender seemed wrong to me here since it can do more than transfer tokens of the owner, it's more generic.

Also worth noting that solmate's implementation actually doesn't use spender: https://github.com/transmissions11/solmate/blob/main/src/tokens/ERC6909.sol#L77 So this seemed like a good alternative to me

Author: But I'm happy to change to spender if you think that's better

Reviewer: seems fine as is
